pkg/pillar/cmd/logmanager/logmanager.go

```diff
@@ -536,81 +532,99 @@ func processEvents(image string, prevLastSent time.Time,
 
 			log.Debugf("processEvents(%s): sending at messageCount %d, byteCount %d\n",
 				image, messageCount, byteCount)
-			sent = sendProtoStrForLogs(reportLogs, image,
-				iteration, eveVersion)
-			messageCount = 0
-			iteration++
-			if sent {
-				recordLast(lastSentDirname, image)
+			if event.isAppLog {
+				sent = sendProtoStrForAppLogs(appUUID, appLogBundle, iteration, image)
+				delete(appLogBundles, appUUID)
 			} else {
-				recordLast(lastDeferDirname, image)
+				sent = sendProtoStrForLogs(reportLogs, image,
+					iteration, eveVersion)
+			}
+			iteration++
+			if !sent {
 				deferInprogress = true
 			}
 
 		case <-flushTimer.C:
-			if messageCount == 0 {
-				break
-			}
 			log.Debugf("processEvents(%s) flush at %s dropped %d messageCount %d bytecount %d\n",
 				image, time.Now().String(),
 				dropped, messageCount,
 				proto.Size(reportLogs))
-			sent := sendProtoStrForLogs(reportLogs, image,
-				iteration, eveVersion)
-			messageCount = 0
+			// Iterate the app log bundle map and send out all app logs
+			sent := flushAllLogBundles(image, iteration, eveVersion, reportLogs, appLogBundles)
 			iteration++
-			if sent {
-				recordLast(lastSentDirname, image)
-			} else {
-				recordLast(lastDeferDirname, image)
+			if !sent {
 				deferInprogress = true
 			}
 		}
 	}
 }
 
-// Touch/create a file to keep track of when things where sent before a reboot
-func recordLast(dirname string, image string) {
-	log.Debugf("recordLast(%s, %s)\n", dirname, image)
-	filename := fmt.Sprintf("%s/%s/%s", types.PersistDir, dirname, image)
-	_, err := os.Stat(filename)
-	if err != nil {
-		file, err := os.Create(filename)
-		if err != nil {
-			log.Infof("recordLast: %s\n", err)
-			return
+func flushAllLogBundles(image string, iteration int, eveVersion string,
+	reportLogs *logs.LogBundle, appLogBundles map[string]*logs.AppInstanceLogBundle) bool {
+	sent := sendProtoStrForLogs(reportLogs, image, iteration, eveVersion)
+	if !sent {
+		return false
+	}
+	appBundlesToDelete := []string{}
+	for appUUID, appLogBundle := range appLogBundles {
+		log.Debugf("flushAllLogBundles: Trying to flush App bundle with UUID %s and %d logs",
+			appUUID, len(appLogBundle.Log))
+		sent = sendProtoStrForAppLogs(appUUID, appLogBundle, iteration, image)
```

Comment:
There could be zero events in this bundle, Should avoid sending in that case.

Reply:
Taken care of inside the callee